File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    void ret(vector<string>&left,vector<string>&v){
        vector<string>a;
        /*for(auto it:left){
            cout<<it<<' ';
        }
        cout<<'\n';*/
        int x=left.size();
        for(int i=0;i<x;i++){
            for(auto it:v){
                a.push_back(left[i]+it);
            }
        }
        left=a;
    }
    vector<string> letterCombinations(string digits) {
        unordered_map<char,vector<string>>m;
        m['2']={"a","b","c"};
        m['3']={"d","e","f"};
        m['4']={"g","h","i"};
        m['5']={"j","k","l"};
        m['6']={"m","n","o"};
        m['7']={"p","q","r","s"};
        m['8']={"t","u","v"};
        m['9']={"w","x","y","z"};
        vector<string>ans;
        for(int i=0;i<digits.size();i++){
            if(i==0){
                ans=m[digits[i]];
            }
            else{
                ret(ans,m[digits[i]]);
            }
        }
        return ans;
    }
};

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if(digits==""):
            return []
        dict={
            '2':["a","b","c"],
            '3':["d","e","f"],
            '4':["g","h","i"],
            '5':["j","k","l"],
            '6':["m","n","o"],
            '7':["p","q","r","s"],
            '8':["t","u","v"],
            '9':["w","x","y","z"]
        }
        intial=dict[digits[0]]
        def ret(val):
            nonlocal intial
            temp=[]
            for i in range(len(intial)):
                for j in range(len(dict[val])):
                    temp.append(intial[i]+dict[val][j])
            intial=temp
        for i in range(1,len(digits),1):
            ret(digits[i])
        return intial